core/network/src/protocol/message.rs

```diff
@@ -284,16 +286,21 @@ pub mod generic {
 			if let Some(state) = &self.state {
 				state.encode_to(dest);
 			}
+			if !self.data.is_empty() {
```

Comment:
This will interfere with encoding any additional fields. I.e. this new protocol version should always encode `data` when communicating with the same protocol version and do not encode when communicating with an older version. I'd suggest using `Option<Vec<u8>>` in `BlockAnnounce` for that. Later we can remove backwards compatibility and `Option` without changing the actual encoding for the current version.

Reply:
I would normally have called `encode_to` unconditionally. Wrapping `data` in an `Option` is not bulletproof either, as one can easily just put it in a `Some`. Also, changing the type ripples through the whole system which is not great. As I mentioned [elsewhere](https://github.com/paritytech/substrate/pull/3452#issuecomment-525172177), we should reconsider our encoding, otherwise minor changes such as this will continue to hurt us.

Reply:
Sure, we should get proper versioning support at some point, but this is out of scope of this PR I believe.